File: typescript/test/messageResolver.test.ts
import { describe, expect, it } from "vitest";
import { MessageResolver } from "../src/messageResolver.js";
import { ValidatorRegistry } from "../src/validators.js";

describe("MessageResolver", () => {
  it("defaults to Japanese (matches the former hardcoded text)", () => {
    const m = new MessageResolver();
    expect(m.resolve("required")).toBe("必須項目です");
    expect(m.resolve("maxLength", { value: 3 })).toBe("3文字以内で入力してください");
  });

  it("falls back to the key for unknown keys", () => {
    expect(new MessageResolver().resolve("nope")).toBe("nope");
  });

  it("supports locale override and switching", () => {
    const m = new MessageResolver({
      locale: "en",
      messages: { en: { required: "Required", maxLength: "Max {value} chars" } },
    });
    expect(m.resolve("required")).toBe("Required");
    expect(m.resolve("maxLength", { value: 3 })).toBe("Max 3 chars");
    // key missing in en falls back to ja
    expect(m.resolve("email")).toBe("メールアドレスの形式が正しくありません");
    // same tables, locale switched back to ja
    expect(m.withLocale("ja").resolve("required")).toBe("必須項目です");
  });
});

describe("ValidatorRegistry i18n", () => {
  it("keeps Japanese by default", () => {
    const r = new ValidatorRegistry();
    expect(r.run("", { type: "required", params: {} })).toBe("必須項目です");
  });

  it("localizes built-in messages via an injected resolver", () => {
    const r = new ValidatorRegistry(
      undefined,
      new MessageResolver({
        locale: "en",
        messages: { en: { required: "Required", maxLength: "Max {value} chars" } },
      }),
    );
    expect(r.run("", { type: "required", params: {} })).toBe("Required");
    expect(r.run("ABCD", { type: "maxLength", params: { value: 3 } })).toBe(
      "Max 3 chars",
    );
  });
});

File: flutter/packages/hatake_core/test/message_resolver_test.dart
import 'package:hatake_core/hatake_core.dart';
import 'package:test/test.dart';

void main() {
  group('MessageResolver', () {
    test('既定は日本語（従来の固定文言と一致）', () {
      final m = MessageResolver();
      expect(m.resolve('required'), '必須項目です');
      expect(m.resolve('maxLength', {'value': 3}), '3文字以内で入力してください');
    });

    test('未知キーはキー名にフォールバック', () {
      expect(MessageResolver().resolve('nope'), 'nope');
    });

    test('ロケール上書き・切替ができる', () {
      final m = MessageResolver(
        locale: 'en',
        messages: {
          'en': {'required': 'Required', 'maxLength': 'Max {value} chars'},
        },
      );
      expect(m.resolve('required'), 'Required');
      expect(m.resolve('maxLength', {'value': 3}), 'Max 3 chars');
      // en に無いキーは ja にフォールバック
      expect(m.resolve('email'), 'メールアドレスの形式が正しくありません');
      // 同じテーブルのままロケールだけ ja に戻す
      expect(m.withLocale('ja').resolve('required'), '必須項目です');
    });
  });

  group('ValidatorRegistry の i18n', () {
    test('既定（日本語）は変わらない', () {
      final r = ValidatorRegistry();
      expect(
        r.run('', const ValidatorDefinition(type: ValidatorTypes.required)),
        '必須項目です',
      );
    });

    test('リゾルバ注入で英語メッセージになる', () {
      final r = ValidatorRegistry(
        null,
        MessageResolver(
          locale: 'en',
          messages: {
            'en': {'required': 'Required', 'maxLength': 'Max {value} chars'},
          },
        ),
      );
      expect(
        r.run('', const ValidatorDefinition(type: ValidatorTypes.required)),
        'Required',
      );
      expect(
        r.run(
          'ABCD',
          const ValidatorDefinition(
              type: ValidatorTypes.maxLength, params: {'value': 3}),
        ),
        'Max 3 chars',
      );
    });

    test('ルール個別の message 上書きはロケールより優先（従来どおり）', () {
      // message 上書きは FormValidator 側で解決されるため、ここでは
      // ValidatorDefinition.message が保持されることだけ確認する。
      const def = ValidatorDefinition(
          type: ValidatorTypes.required, message: 'カスタム');
      expect(def.message, 'カスタム');
    });
  });
}
